python/lsst/ci/prepare.py

```diff
@@ -140,16 +141,32 @@ class ProductFetcher(object):
         :ivar refs: A list of refs to attempt to git-checkout
         :ivar no_fetch: If true, don't fetch, just checkout the first matching ref.
     """
-    def __init__(self, build_dir, repository_patterns, refs, no_fetch):
+    def __init__(self, build_dir, repos, repository_patterns, refs, no_fetch):
         self.build_dir = os.path.abspath(build_dir)
         self.refs = refs
-        self.repository_patterns = repository_patterns.split('|')
+        if repository_patterns:
+            self.repository_patterns = repository_patterns.split('|')
+        else:
+            self.repository_patterns = None
         self.no_fetch = no_fetch
+        if repos:
+            if os.path.exists(repos):
+                f = open(repos, 'r')
+                self.repos = yaml.safe_load(f)
+            else:
+                raise Exception("YAML repos file '%s' does not exist" % repos)
+        else:
+            self.repos = None
 
     def _origin_candidates(self, product):
         """ Expand repository_patterns into URLs. """
         data = { 'product': product }
-        return [ pat % data for pat in self.repository_patterns ]
+        locations = []
+        if self.repos and product in self.repos:
+            locations.append(self.repos[product])
```

Comment:
So `--repos` takes precedence over `REPOSITORY_PATTERN`?  Should probably be documented somewhere.  I thought you were just going to map product names to repository names but still use the `REPOSITORY_PATTERN` to turn them into URLs.  Going the whole way in the YAML is fine, though.